compiler/rustc_middle/src/ty/consts/valtree.rs

```diff
@@ -20,6 +22,7 @@ pub enum ValTree<'tcx> {
     /// See the `ScalarInt` documentation for how `ScalarInt` guarantees that equal values
     /// of these types have the same representation.
     Leaf(ScalarInt),
+    SliceOrStr(ValSlice<'tcx>),
```

Comment:
I am not too happy about that variant as it will make equality between valtrees more complex, especially if we end up changing `Leaf` to `&'tcx ty::Const<'tcx>` instead of `ScalarInt`, see https://github.com/rust-lang/project-const-generics/issues/43 for more details.

I am however worried that `struct WithLongText<const TEXT: &'static str>` will actually be used somewhat frequently, so it's very much possible that we can't avoid this variant. I think I slightly prefer to keep using a lot of `Leaf`s for that until we get a complaint by users. I do want @oli-obk's input here

Reply:
yea, we should make switching to `ValSlice` a separate PR from the rest of the work so we can benchmark it once the logic works.

Reply:
FWIW I think the way to avoid making equality complicated is to ensure there is a normal form: *all* (parts of( constants of type `&str` (for example) must be a `SliceOrStr`, they can *never* be a lot of `Leaf`s. That ensures that when comparing valtrees of equal type, we never have ambiguity about whether something should be represented as `Leaf` or `SliceOrStr`.